kartothek/io/dask/delayed.py

```diff
@@ -582,3 +585,45 @@ def store_delayed_as_dataset(
         metadata_merger=metadata_merger,
         metadata_storage_format=metadata_storage_format,
     )
+
+
+@default_docs
+def build_dataset_indices__delayed(
+    store, dataset_uuid, columns, partition_size=None, factory=None
+):
+    """
+    Function which builds a :class:`~kartothek.core.index.ExplicitSecondaryIndex`.
+
+    This function loads the dataset, computes the requested indices and writes
+    the indices to the dataset. The dataset partitions itself are not mutated.
+
+    Parameters
+    ----------
+    partition_size: Optional[int]
+        Dask bag partition size. Use a larger numbers to decrease scheduler load and overhead, use smaller numbers for a
+        fine-grained scheduling and better resilience against worker errors.
+
+    Returns
+    -------
+    A dask.delayed computation object.
+    """
+    ds_factory = _ensure_factory(
+        dataset_uuid=dataset_uuid,
+        store=store,
+        factory=factory,
+        load_dataset_metadata=False,
+    )
+
+    mps = read_dataset_as_metapartitions(factory=ds_factory)
+
+    return (
+        db.from_sequence(seq=mps, partition_size=partition_size)
+        .map(MetaPartition.build_indices, columns=columns)
+        .map(MetaPartition.remove_dataframes)
+        .reduction(list, list, split_every=False, out_type=db.Bag)
+        .flatten()
+        .map_partitions(list)
+        .map_partitions(
```

Comment:
Isn't the output type of this function a dask.Bag, i.e. for a true delayed interface we should call `to_delayed`? I aggree with you that this is an implementation detail but considering the that we distinguish these modules by what kind of dask collection they return we could offer both functions (where the delayed version simply calls build_dataset_indices_bag().to_delayed)

Reply:
Then I would move this to the our bag submodule entirely because bags use a different optimizer which can make a huge difference, especially for the reduction and the map calls.

Reply:
Well if you call `build_dataset_indices_bag().to_delayed(optimize=False)` you fall back to the unoptimized / pure delayed graph. My point is simply that since we introduced these modules we should make use of them.

Reply:
I've moved it.